KeenClient/KIODBStore.m

```diff
@@ -143,35 +157,45 @@ - (BOOL)deleteAndRecreateCorruptDB {
 }
 
 - (BOOL)openAndInitDB {
-    if(!dbIsOpen) {
-        if (![self openDB]) {
-            return false;
-        } else {
-            if(![self createTables]) {
-                KCLogError(@"Failed to create SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self migrateTable]) {
-                KCLogError(@"Failed to migrate SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self prepareAllSQLiteStatements]) {
+
+    // The database can be opened from the thread calling into the SDK,
+    // or by the database queue thread if it has been closed, so open
+    // needs to be protected with a lock.
+    [openLock lock];
```

Comment:
NSLock uses a normal POSIX mutex under the hood. Another alternative could be a spin lock if we decide/find there will be high contention, or an atomic swap and compare could guard the mutex lock.

Most calls to KIODBStore methods seem to make a single call to checkOpenDB before dispatching a block to the queue, while the queue sometimes calls a helper that will call checkOpenDB. I would expect that taking the lock wouldn't really be a huge bottleneck unless there were a pattern of contention or a massive volume of events, which doesn't seem likely either.

For example, calling addEvent: 1000 times will take and free the lock 1000 times, and the dispatched block doesn't take a lock. [This](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/Multithreading/ThreadSafety/ThreadSafety.html) says an atomic operation will be about 4x faster than taking a mutex. That documentation gives some timing examples for pretty old hardware, but maybe it could compare to a modern phone. Taking a mutex would take 0.2 us on the hardware specified vs 0.05 us for an atomic swap. So let's say that means locking overhead is `1000 * 0.5 us * 2 = 1000 us = 1 ms`, possibly with a longer wait in there if there are a few operations already queued up that actually take the lock thus causing contention. In terms of perf, I'd be more concerned about, say, writing those events to disk.

Let me know if you think that's a reasonable take on the situation or think a spin lock or atomic swap is worth it (could do an atomic compare and only take the lock if the db isn't already open).

In general though, `KIODBStore` is kind of a can of worms. If time allows, I would rip things up and ensure no lock is needed, and that we're not closing the database unless it needs to be completely torched, thus not needing to check if it's open in order to re-open it.

There's kind of a pile of stuff surrounding the data store that should be dealt with at some point, ideally it would be for `vnext_major`: #78, #170, and #183 kind of all go in that bucket. I'd say the work would include #78, and cleaning up error handling in `KIODBStore` to ensure it isn't closed unless the store is going to be erased, and ensuring opens only happen from a single thread, likely the dispatch queue that handles db operations.